mountpoint-s3/src/inode.rs

```diff
@@ -131,55 +131,63 @@ impl Superblock {
     /// The kernel may forget a number of references (`n`) in one forget message to our FUSE implementation.
     /// If the lookup count reaches zero, it is safe for the [Superblock] to delete the [Inode].
     pub fn forget(&self, ino: InodeNo, n: u64) {
-        let mut inodes = self.inner.inodes.write().unwrap();
-        match inodes.get(&ino) {
-            None => {
-                debug_assert!(
-                    false,
-                    "forget should not be called on inode already removed from superblock"
-                );
-                error!("forget called on inode {ino} already removed from the superblock");
-            }
-            Some(inode) => {
-                logging::record_name(inode.name());
-                let new_lookup_count = inode.dec_lookup_count(n);
-                if new_lookup_count == 0 {
-                    // Safe to remove, kernel no longer has a reference to it.
-                    trace!(ino, "removing inode from superblock");
-                    let inode = inodes.remove(&ino).expect("we just retrieved it");
-
-                    // Should be impossible for this to fail (VFS inodes reference their parent, so
-                    // children need to be freed first), but let's not crash in a `forget` function...
-                    let Some(parent) = inodes.get(&inode.parent()) else {
-                        debug_assert!(false, "children should be forgotten before parents");
+        // Put inode removal in a block so we don't hold the lock on the inodes table longer than needed.
+        let (inode, parent) = {
+            let mut inodes = self.inner.inodes.write().unwrap();
```

Comment:
We should avoid holding a write lock on the Inodes map for long in the first place. Especially while we need to lock other mutexes, like in `dec_lookup_count`.

What if we took just a read lock here (and released immediately), and then, only if required, took a write lock before line 152?